### PR TITLE
Improve error messaging on invalid coordinates.

### PR DESCRIPTION
**Pull request by tjstienstra**
#### Brief description of what is fixed or changed
Improve error messaging on invalid coordinates by implementing a `_validate_coordinate` function.

#### Release Notes

* physics.mechanics
  * Improve error messaging on invalid coordinates.

### CODE REVIEW

**Comment by TJStienstra** [comment]
This is the PR mentioned in [#24188 comment](https://github.com/sympy/sympy/pull/24188#issuecomment-1294839757).

**Comment by TJStienstra** [comment]
@brocksam @moorepants FYI

File: sympy/physics/mechanics/tests/test_kane.py [hunk 1]
```diff
@@ -9,6 +9,20 @@
...
+    l, m, g = symbols('l m g')
+    q, u = symbols('q u')  # Generalized coordinate
+    kd = [q.diff(dynamicsymbols._t) - u]
+    N, O = ReferenceFrame('N'), Point('O')
+    O.set_vel(N, 0)
+    P = Particle('P', Point('P'), m)
+    P.point.set_pos(O, l * (sin(q) * N.x - cos(q) * N.y))
+    F = (P.point, -m * g * N.y)
+    raises(ValueError, lambda: KanesMethod(N, [q], [u], kd, bodies=[P],
+                                           forcelist=[F]))
```
Comment (TJStienstra) [inline sympy/physics/mechanics/tests/test_kane.py:23]: This test would also pass, when the `_validate_coordinates()` call in `KanesMethod.__init__` would be removed (only the error message is now a bit more clear).

File: sympy/physics/mechanics/tests/test_lagrange.py [hunk 2]
```diff
@@ -7,6 +7,20 @@
...
+    # Simple pendulum, but use symbol instead of dynamicsymbol
+    l, m, g = symbols('l m g')
+    q = symbols('q')  # Generalized coordinate
+    N, O = ReferenceFrame('N'), Point('O')
+    O.set_vel(N, 0)
+    P = Particle('P', Point('P'), m)
+    P.point.set_pos(O, l * (sin(q) * N.x - cos(q) * N.y))
+    P.potential_energy = m * g * P.point.pos_from(O).dot(N.y)
+    L = Lagrangian(N, P)
+    raises(ValueError, lambda: LagrangesMethod(L, [q], bodies=P))
```
Comment (TJStienstra) [inline sympy/physics/mechanics/tests/test_lagrange.py:23]: Would normally have crashed when forming the lagrange equations.

**Comment by TJStienstra** [comment]
Should `JointsMethod` also call `_validate_coordinates`? Seems a bit redundant to me.

File: sympy/physics/mechanics/functions.py [hunk 3]
```diff
@@ -729,3 +729,33 @@ def flist_iter():
...
+        coordinates = [coordinates]
+    if speeds is None:
+        speeds = []
+    elif not iterable(speeds):
+        speeds = [speeds]
+
+    if check_duplicates:  # Check for duplicates
+        if len(coordinates) != len(set(coordinates)):
+            raise ValueError('Duplicate generalized coordinates found, all '
+                             'generalized coordinates should be unique.')
```
Comment (brocksam) [inline sympy/physics/mechanics/functions.py:749]: If the purpose of this PR is to improve error messaging, do we want to be even more descriptive for the user here? I.e. by specifying which coordinate associated with which joint is duplicated? Or are we happy that this is sufficient? Same goes for the generalised speeds error message below.

**Review by brocksam — APPROVED** [review]
This is a nice improvement, well done @TJStienstra.

File: sympy/physics/mechanics/functions.py (hunk 3 above)
Comment (TJStienstra) [reply sympy/physics/mechanics/functions.py:749]: Yeah, in a way I wanted to do so, but if this is preferred we have to create a loop. And the user can quite easily check the coordinates and speeds parsed to the function.

**Pull request merged**